File: skills/transx-translate/scripts/history.py
"""Write translation history in the same format used by TransX CLI."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
import json
import os
from pathlib import Path
import tempfile
import time
import uuid


HISTORY_DIR = Path.home() / ".transx" / "history"
LOCK_PATH = HISTORY_DIR / ".lock"


def china_timestamp() -> str:
    current = datetime.now(timezone.utc) + timedelta(hours=8)
    return current.strftime("%Y-%m-%d %H:%M:%S.") + f"{current.microsecond // 1000:03d}"


def write_json_atomic(path: Path, value: object) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temporary = tempfile.mkstemp(prefix=f".{path.name}.", dir=path.parent)
    try:
        with os.fdopen(fd, "w", encoding="utf-8", newline="\n") as handle:
            json.dump(value, handle, ensure_ascii=False, indent=2)
            handle.write("\n")
        if os.name != "nt":
            os.chmod(temporary, 0o600)
        os.replace(temporary, path)
    except BaseException:
        try:
            os.unlink(temporary)
        except FileNotFoundError:
            pass
        raise


def acquire_lock() -> int:
    HISTORY_DIR.mkdir(parents=True, exist_ok=True)
    deadline = time.monotonic() + 5
    while time.monotonic() < deadline:
        try:
            return os.open(LOCK_PATH, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
        except FileExistsError:
            try:
                if time.time() - LOCK_PATH.stat().st_mtime > 30:
                    LOCK_PATH.unlink(missing_ok=True)
                    continue
            except FileNotFoundError:
                continue
            time.sleep(0.05)
    raise RuntimeError("翻译历史正在被其他进程使用，请稍后重试")


def rebuild_index(updated_at: str, last_warning_at: str | None) -> dict[str, object]:
    records: list[dict[str, object]] = []
    total_bytes = 0
    for file_path in sorted(HISTORY_DIR.glob("????-??-??.json")):
        raw = file_path.read_bytes()
        total_bytes += len(raw)
        parsed = json.loads(raw.decode("utf-8"))
        if isinstance(parsed, dict) and isinstance(parsed.get("records"), list):
            records.extend(parsed["records"])
    timestamps = sorted(
        record["createdAt"]
        for record in records
        if isinstance(record, dict) and isinstance(record.get("createdAt"), str)
    )
    return {
        "version": 1,
        "updatedAt": updated_at,
        "totalRecords": len(records),
        "totalBytes": total_bytes,
        "oldestAt": timestamps[0] if timestamps else None,
        "newestAt": timestamps[-1] if timestamps else None,
        "lastWarningAt": last_warning_at,
    }


def append_history(
    source_lang: str,
    target_lang: str,
    input_text: str | None = None,
    output_text: str | None = None,
    *,
    source_file_path: str | None = None,
    output_file_path: str | None = None,
) -> None:
    lock_fd = acquire_lock()
    try:
        created_at = china_timestamp()
        date = created_at[:10]
        daily_path = HISTORY_DIR / f"{date}.json"
        if daily_path.exists():
            daily = json.loads(daily_path.read_text(encoding="utf-8"))
            if not isinstance(daily, dict) or not isinstance(daily.get("records"), list):
                raise RuntimeError(f"翻译历史文件无效：{daily_path.name}")
        else:
            daily = {"version": 1, "date": date, "records": []}
        record: dict[str, object] = {
            "id": str(uuid.uuid4()),
            "createdAt": created_at,
            "sourceLang": source_lang,
            "targetLang": target_lang,
        }
        if source_file_path is not None:
            source_path = Path(source_file_path)
            output_path = Path(output_file_path) if output_file_path else None
            record.update(
                {
                    "format": "file",
                    "sourceFilePath": str(source_path),
                    "sourceFileName": source_path.name,
                    "outputFilePath": str(output_path) if output_path else None,
                    "outputFileName": output_path.name if output_path else None,
                }
            )
        else:
            record.update(
                {
                    "format": "plain",
                    "input": input_text or "",
                    "output": output_text or "",
                }
            )
        daily["records"].append(record)
        write_json_atomic(daily_path, daily)
        index_path = HISTORY_DIR / "index.json"
        last_warning_at = None
        if index_path.exists():
            try:
                previous = json.loads(index_path.read_text(encoding="utf-8"))
                last_warning_at = previous.get("lastWarningAt") if isinstance(previous, dict) else None
            except (OSError, json.JSONDecodeError):
                pass
        write_json_atomic(index_path, rebuild_index(created_at, last_warning_at))
    finally:
        os.close(lock_fd)
        LOCK_PATH.unlink(missing_ok=True)

File: skills/transx-translate/scripts/history.mjs
import { open, mkdir, readFile, readdir, rename, rm, stat, writeFile } from "node:fs/promises";
import { randomUUID } from "node:crypto";
import os from "node:os";
import path from "node:path";


const historyDirectory = path.join(os.homedir(), ".transx", "history");
const lockPath = path.join(historyDirectory, ".lock");

function chinaTimestamp() {
  return new Date(Date.now() + 8 * 60 * 60 * 1000).toISOString().slice(0, -1).replace("T", " ");
}

async function writeJsonAtomic(filePath, value) {
  await mkdir(path.dirname(filePath), { recursive: true, mode: 0o700 });
  const temporary = `${filePath}.${randomUUID()}.tmp`;
  try {
    await writeFile(temporary, `${JSON.stringify(value, null, 2)}\n`, { encoding: "utf8", mode: 0o600 });
    await rename(temporary, filePath);
  } catch (error) {
    await rm(temporary, { force: true });
    throw error;
  }
}

async function acquireLock() {
  await mkdir(historyDirectory, { recursive: true, mode: 0o700 });
  const deadline = Date.now() + 5000;
  while (Date.now() < deadline) {
    try {
      return await open(lockPath, "wx", 0o600);
    } catch (error) {
      if (error?.code !== "EEXIST") throw error;
      try {
        const current = await stat(lockPath);
        if (Date.now() - current.mtimeMs > 30000) {
          const stale = `${lockPath}.${randomUUID()}.stale`;
          await rename(lockPath, stale);
          await rm(stale, { force: true });
          continue;
        }
      } catch (statError) {
        if (statError?.code !== "ENOENT") throw statError;
      }
      await new Promise((resolve) => setTimeout(resolve, 50));
    }
  }
  throw new Error("翻译历史正在被其他进程使用，请稍后重试");
}

async function listDailyFiles() {
  try {
    return (await readdir(historyDirectory)).filter((name) => /^\d{4}-\d{2}-\d{2}\.json$/.test(name)).sort();
  } catch (error) {
    if (error?.code === "ENOENT") return [];
    throw error;
  }
}

async function rebuildIndex(updatedAt, lastWarningAt) {
  const records = [];
  let totalBytes = 0;
  for (const fileName of await listDailyFiles()) {
    const raw = await readFile(path.join(historyDirectory, fileName), "utf8");
    totalBytes += Buffer.byteLength(raw);
    const parsed = JSON.parse(raw);
    if (Array.isArray(parsed?.records)) records.push(...parsed.records);
  }
  const timestamps = records
    .map((record) => record?.createdAt)
    .filter((value) => typeof value === "string")
    .sort();
  return {
    version: 1,
    updatedAt,
    totalRecords: records.length,
    totalBytes,
    oldestAt: timestamps[0] ?? null,
    newestAt: timestamps.at(-1) ?? null,
    lastWarningAt,
  };
}

export async function appendHistory(recordInput) {
  const lock = await acquireLock();
  try {
    const createdAt = chinaTimestamp();
    const date = createdAt.slice(0, 10);
    const dailyPath = path.join(historyDirectory, `${date}.json`);
    let daily;
    try {
      daily = JSON.parse(await readFile(dailyPath, "utf8"));
      if (!Array.isArray(daily?.records)) throw new Error(`翻译历史文件无效：${date}.json`);
    } catch (error) {
      if (error?.code !== "ENOENT") throw error;
      daily = { version: 1, date, records: [] };
    }
    daily.records.push({
      id: randomUUID(),
      createdAt,
      sourceLang: recordInput.sourceLang,
      targetLang: recordInput.targetLang,
      ...(recordInput.format === "file"
        ? {
            format: "file",
            sourceFilePath: recordInput.sourceFilePath,
            sourceFileName: recordInput.sourceFileName,
            outputFilePath: recordInput.outputFilePath,
            outputFileName: recordInput.outputFileName,
          }
        : { format: "plain", input: recordInput.input, output: recordInput.output }),
    });
    await writeJsonAtomic(dailyPath, daily);
    const indexPath = path.join(historyDirectory, "index.json");
    let lastWarningAt = null;
    try {
      lastWarningAt = JSON.parse(await readFile(indexPath, "utf8"))?.lastWarningAt ?? null;
    } catch (error) {
      if (error?.code !== "ENOENT" && !(error instanceof SyntaxError)) throw error;
    }
    await writeJsonAtomic(indexPath, await rebuildIndex(createdAt, lastWarningAt));
  } finally {
    await lock.close();
    await rm(lockPath, { force: true });
  }
}
